### ISSUE TITLE
React-select not scrolling to the selected value on menu re-open

### ISSUE THREAD

**Issue by ColbySong** [JedWatson/react-select#3648]
**react-select: 3.0.4**

After selecting a value and then re-opening the menu, the component does not scroll the selected value into view.

This behaviour occurs for both standard options and also the new options with group headers.

Options with group header issue:
https://codesandbox.io/s/p68rh?module=/example.js
- select an option far down the list
- re-open menu, selected element is highlighted but not scrolled into view when menu is opened

Standard options:
https://codesandbox.io/s/kwv9t?module=/example.js
- select option 'silver' which is far down the list
- re-open menu and you will see the view does not scroll to the selected value

similar issue reported by #3535 for custom options as well

**Comment by ankush29** [comment]
solved https://github.com/JedWatson/react-select/pull/3652/commits/bb9c1d7bf53a68aa6bd4e6be5c40c3b0c14a7c01

**Comment by csantos1113** [comment]
Same issue as https://github.com/JedWatson/react-select/issues/3656

**Comment by csantos1113** [comment]
Thanks @ankush29 - will be fixed on PR https://github.com/JedWatson/react-select/pull/3652

**Comment by imnikij** [comment]
Hey @csantos1113 I see this issue as still an open one, any idea by when this will be fixed ?

**Comment by tonistenno** [comment]
Really waiting for this fix

**Comment by alexmnv** [comment]
Had the same issue. The problem turned out to be that the option object passed to `value` prop was not  the same object present in `options` array, and `react-select` uses strict equality comparison (options.indexOf) to find selected option. E.g.:
```
<Select
  options={options}
  
  // wrong: new object is created. `options.indexOf(value)` fails to find the option index
  value={{
	id: selectedId,
	name: selectedName
  }}
/>

<Select
  options={options}
  
  // correct: `value` can be found in `options` array
  value={options[selectedIndex]}
/>
```

**Comment by milosh86** [comment]
This issue is introduced in 3.0.4 release, probably this feature https://github.com/JedWatson/react-select/issues/3648 cause it.

**Comment by Hodor9898** [comment]
I am having the same issue

**Comment by TBear79** [comment]
Yes. I also have the same issue. I'm objects with same reference as value but it doesn't seem to be working. When is this going to be fixed?

**Comment by AnthonyCrowcroft** [comment]
just rolled back to version 3.0.3 and it seems to still have the issue

**Comment by AnthonyCrowcroft** [comment]
any expected date for a fix for this looking back at 3.0.3 bug still seems to be present for me.

**Comment by DominicFath** [comment]
We are on version 3.0.8 and this bug is still occurring. Would be much appreciated if this could be solved as quickly as possible since it is quite an obvious front facing issue.

**Comment by carlosriveroib** [comment]
Please, any updates on this?

**Comment by konsvarl** [comment]
Guys, any updates ? +

**Comment by vthai** [comment]
Is this going to be fixed in the next release?

**Comment by maidinhkhoa92** [comment]
I tried my self to fix issue.
Following this commit.
> solved [bb9c1d7](https://github.com/JedWatson/react-select/commit/bb9c1d7bf53a68aa6bd4e6be5c40c3b0c14a7c01)

in src/select.js
openMenu functionality:
`openMenu(focusOption: 'first' | 'last') {
    const { menuOptions, selectValue, isFocused } = this.state;
    const { isMulti } = this.props;
    let openAtIndex =
      focusOption === 'first' ? 0 : menuOptions.focusable.length - 1;

    if (!isMulti) {
      const selectedIndex = this.props.options.indexOf(selectValue[0]);
      if (selectedIndex > -1) {
        openAtIndex = selectedIndex;
      }
    }

    // only scroll if the menu isn't already open
    this.scrollToFocusedOptionOnUpdate = !(isFocused && this.menuListRef);
    this.inputIsHiddenAfterUpdate = false;
    this.onMenuOpen();
    this.setState({
      focusedValue: null,
      focusedOption: this.props.options[openAtIndex],
    });

    this.announceAriaLiveContext({ event: 'menu' });
  }`

getNextFocusedOption functionality:
`getNextFocusedOption(options: OptionsType) {
    const { selectValue: lastFocusedOption } = this.state;
    const { options : menuOptions } = this.props;
    return lastFocusedOption && menuOptions.indexOf(lastFocusedOption[0]) > -1
      ? lastFocusedOption[0]
      : options[0];
  }`

**Comment by matyas-igor** [comment]
Worked for me pretty well with defining custom `onMenuOpen` prop:

```jsx
import React, { useCallback, useRef } from 'react'
import Select from 'react-select'

const MySelect = ({ isMulti, onMenuOpen: onMenuOpenProp, ...props }) => {
  const selectRef = useRef(null)

  // Feature of focusing selected option when menu is getting opened
  const onMenuOpen = useCallback(() => {
    // Calling an initial prop if it has been passed
    onMenuOpenProp?.()
    if (!isMulti) {
      // Getting a selected option
      const option = selectRef.current?.select?.state?.selectValue?.[0]
      if (option) {
        setTimeout(() => {
          // Setting a focused value as a selected one
          // Reference: https://github.com/JedWatson/react-select/blob/master/packages/react-select/src/Select.js#L503
          if (selectRef.current?.select) {
            selectRef.current.select.scrollToFocusedOptionOnUpdate = true
            selectRef.current.select.inputIsHiddenAfterUpdate = false
            selectRef.current.select.setState({
              focusedValue: option?.value || null,
              focusedOption: option,
            })
          }
        })
      }
    }
  }, [selectRef.current, onMenuOpenProp])

  return <Select ref={selectRef} isMulti={isMulti} onMenuOpen={onMenuOpen} {...props} />
}
```

**Comment by soulmerge** [comment]
Can confirm that the above work-around works in version 3.0.8.
Here are canonical versions of the URLs in the comments of the snippet above:

https://github.com/JedWatson/react-select/blob/d78730377e25658a5df2930faee54574e757d463/packages/react-select/src/Select.js#L503-L506

https://github.com/JedWatson/react-select/blob/d78730377e25658a5df2930faee54574e757d463/packages/react-select/src/Select.js#L802

**Comment by nielskrijger** [comment]
This issue was fixed for me in 3.1.0

**Comment by bladey** [comment]
Hello -

In an effort to sustain the `react-select` project going forward, we're closing old issues.

We understand this might be inconvenient but in the best interest of supporting the broader community we have to direct our efforts towards the current major version.

If you aren't using the latest version of `react-select` please consider upgrading to see if it resolves any issues you're having.

However, if you feel this issue  is still relevant and you'd like us to review it - please leave a comment and we'll do our best to get back to you!

**Comment by peterp** [comment]
@bladey This should be re-opened.

**Comment by bladey** [comment]
Thanks @peterp.

**Comment by ebonow** [comment]
Greetings, I can replicate the issue as well with the following criteria with or without custom components. 
``` JS
<Select 
isMulti 
hideSelectedOptions={false} 
options={groups} 
/>
```
https://codesandbox.io/s/react-select-bug-scroll-selected-group-option-k9942?file=/src/App.js

**Comment by louisptremblay** [comment]
Adding a setTimeout to /node_modules/react-select/dist/Select-#####.browser.esm.js seems to fix it for me (4.1.0)
ie. change:
```
scrollIntoView(this.menuListRef, this.focusedOptionRef);
```
to:
```
 setTimeout(() => scrollIntoView(this.menuListRef, this.focusedOptionRef), 0);
```

**Comment by raphaelaleixo** [comment]
Any idea if this bug will be fixed soon? @louisptremblay , have you submitted a pull request with your fix?

**Comment by louisptremblay** [comment]
PR submitted... you can use yarn patch-package in the meantime

**Comment by mhammadsiddiqui** [comment]
So I was able to solve this using onMenuOpen() call back function with timeOut() function as we do not have "MyDropdown__option--is-selected" class at the time when this function is invoked.

My solution is as follows;

```
//when dropdown is not manually clicked then it does not scroll into view
onMenuOpen = () => {
  setTimeout(()=>{
    const selectedEl = document.getElementsByClassName("MyDropdown__option--is-selected")[0];
    if(selectedEl){
      selectedEl.scrollIntoView({behavior:'smooth', block:'nearest', inline: 'start'});
    }
  },15);
};

render(){
const {defaultOption, options} = this.props;

  return (
    <Select 
      options={options}
      value={defaultOption}
      onMenuOpen={this.onMenuOpen}
      onChange={(item: IDropdownOptions) => this.props.onSelect(item)}
      >
    </Select>
  );
}
```

**Comment by ebonow** [comment]
Greetings all, 

Merge conflicts have been resolved and version 4.3 has been released as we wanted to squeeze in at least one more release before the conversion to TypeScript. 

One of the changes that was added was that the focusedOption is now passed as a prop to the MenuList component so hopefully this makes it easier to find the option to be scrolled to if using a custom component. This does not resolve the core issue, but may give some ideas for work arounds for the time being. 

More to come, but wanted to at least give an update that there are eyes on this.

**Comment by railsonsousa106** [comment]
Doesn't work at all for me on version 4.3!
Can you please provide me any code snippet that works?

**Comment by argie2099** [comment]
It's still not working for me, any solution?

**Comment by slaurent22** [comment]
Not working for me on 5.0.0. Are there any code snippets for workarounds?

**Comment by mhammadsiddiqui** [comment]
Here is my the solution, what worked for me : https://stackoverflow.com/a/66710895/7844963

**Comment by slaurent22** [comment]
> 
> 
> Here is my the solution, what worked for me : https://stackoverflow.com/a/66710895/7844963

Thanks so much! I had tried mhammadsiddiqui's solution above, which is similar, but omits the `className` and `classNamePrefix` part

**Comment by StothekMonster** [comment]
Is a fix for this issue scheduled, it's 4 years old?
